### PR TITLE
Homework Analytical Patterns

### PR DESCRIPTION
**Pull request by erich-hs**
My submission for the Analytical Patterns homework.

I've chosen queries 1 to 5 to submit.

For query 5, the instructions state that it should be derived from the OLAP cube built from query 2, but this question does not require results from a GROUPING SET, so I simplified it to answer the question.

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly implemented the state change tracking for `nba_players` using a `CASE` statement to determine the player's state based on their activity and the number of seasons they have played. 

However, there are a few areas where your solution could be improved:

1) **Covering all aspects of the homework prompt:** Your solution does not create a new table for state change tracking as suggested in the example solution. While this is not strictly necessary, it can be beneficial for maintaining a historical record of player states. You might consider adding a `CREATE TABLE` statement at the beginning of your query to store the results.

2) **Readability:** Your code is generally well-formatted and easy to read. However, the comments explaining each condition in the `CASE` statement could be more concise. For example, instead of "Currently active but no previous seasons are 'New' players", you could write "New players are active with no previous seasons". This makes it clear that the condition is defining 'New' players.

3) **Robustness and error handling:** Your solution assumes that the `seasons` array is sorted in ascending order, which may not always be the case. If the array is not sorted, `years_since_last_active` may not accurately reflect the number of years since the player was last active. To handle this, you could sort the `seasons` array before calculating `years_since_last_active`.

4) **Modularity and reusability:** Your solution is fairly modular, with the player state calculation encapsulated in a single `CASE` statement. However, the calculation of `years_since_last_active` is not explicitly shown in your code. If this calculation is complex, you might consider moving it to a separate subquery or CTE for clarity and reusability.

Here's how you might revise your solution based on these suggestions:

```sql
CREATE TABLE nba_players_state_tracking AS
SELECT
  player_name,
  seasons,
  is_active,
  current_season - MAX(seasons) AS years_since_last_active,
  current_season,
  CASE
    WHEN is_active AND CARDINALITY(seasons) = 1 THEN 'New'
    WHEN NOT is_active AND current_season - MAX(seasons) = 1 THEN 'Retired'
    WHEN is_active AND CARDINALITY(seasons) > 1 AND current_season - MAX(seasons) = 0 THEN 'Continued Playing'
    WHEN is_active AND current_season - MAX(seasons) > 1 THEN 'Returned from Retirement'
    WHEN NOT is_active AND current_season - MAX(seasons) > 1 THEN 'Stayed Retired'
  END AS player_state
FROM bootcamp.nba_players
GROUP BY player_name, seasons, is_active, current_season;
```

This revised solution creates a new table `nba_players_state_tracking` and calculates `years_since_last_active` within the query, ensuring that the `seasons` array is sorted by using the `MAX` function.

### Feedback for `query_2.sql`

Overall, your code is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to perform aggregations along the required dimensions. However, there are a few areas where your solution could be improved to better align with the assignment prompt and improve readability and robustness.

1) **Alignment with the assignment prompt:** The assignment prompt asks for aggregations on 'player and team', 'player and season', and 'team'. However, in your solution, you are grouping by 'player_id and team_id', 'player_id and season', and 'team_id'. While this might produce the correct results, it would be more appropriate to use 'player_name' and 'team_abbreviation' as these are more human-readable and align better with the prompt.

2) **Readability:** Your code is generally well-structured and includes helpful comments. However, the use of 'team_id' and 'player_id' in the final output could be confusing, as these are not as easily interpretable as 'team_abbreviation' and 'player_name'. Consider replacing these with the more human-readable versions.

3) **Robustness and error handling:** Your code does not handle the case where a team is the visitor team. In the example solution, a `CASE` statement is used to correctly assign the 'won' value based on whether the team was the home or visitor team. Consider adding similar logic to your solution to ensure it handles all possible scenarios.

4) **Modularity and reusability:** Your code is well-structured and modular, with each step of the process clearly defined in separate CTEs. This makes it easy to understand and modify if needed. However, the final output includes 'team_id' and 'player_id', which may not be useful in all contexts. Consider removing these from the final output to increase the reusability of your code.

Here's a suggestion on how you could modify your code:

```sql
WITH nba_game_details_deduped AS (
  SELECT
    game_id,
    team_abbreviation,
    player_name,
    pts,
    ROW_NUMBER() OVER (PARTITION BY game_id, team_abbreviation, player_name ORDER BY pts DESC) AS row_num
  FROM bootcamp.nba_game_details
),
combined AS (
  SELECT
    gd.game_id,
    gd.team_abbreviation,
    gd.player_name,
    gd.pts,
    g.season
  FROM nba_game_details_deduped gd
    JOIN bootcamp.nba_games g ON gd.game_id = g.game_id
  WHERE
    gd.row_num = 1
),
olap_cube AS (
  SELECT
    COALESCE(team_abbreviation, '(overall)') AS team,
    COALESCE(player_name, '(overall)') AS player,
    COALESCE(CAST(season AS VARCHAR), '(overall)') AS season,
    AVG(pts) AS avg_pts,
    SUM(pts) AS sum_pts
  FROM combined
  GROUP BY GROUPING SETS (
    (player_name, team_abbreviation),
    (player_name, season),
    (team_abbreviation)
  )
)
SELECT
  *
FROM
  olap_cube
```

Keep up the good work! Your understanding of SQL and data manipulation is clear from your solution. With a few tweaks, your code will be even better.

### Feedback for `query_3.sql`

Overall, your code is well-structured and demonstrates a good understanding of SQL and the use of `GROUPING SETS`. However, there are a few areas where it could be improved to better align with the assignment prompt and to enhance readability and efficiency.

1) **Alignment with the assignment prompt:** The prompt asks for a query that identifies the player who scored the most points playing for a single team. Your solution, however, calculates the average points (`avg_pts`) and the sum of points (`sum_pts`) for each player for each season and for each team. While this provides a lot of information, it doesn't directly answer the question. The example solution provides a more direct answer by simply ordering the players by their total points and selecting the top one.

2) **Readability:** Your code is generally well-structured and includes helpful comments, which is great. However, the use of CTEs (Common Table Expressions) and the `GROUPING SETS` clause makes it a bit complex. For this particular problem, a simpler approach like the one in the example solution might be more readable.

3) **Efficiency:** Your solution involves several steps, including deduplicating the `nba_game_details` table, joining it with the `nba_games` table, and then performing a `GROUPING SETS` operation. While these steps are logically correct, they might be unnecessary for this problem and could potentially slow down the query, especially if the tables are large. The example solution is more efficient as it directly queries the aggregated data.

4) **Error handling and robustness:** Your code does not seem to handle potential errors or edge cases. For example, it assumes that the `pts` column does not contain any NULL values. It would be good to add some error handling or checks for this.

5) **Modularity and reusability:** Your code is modular and the use of CTEs makes it easy to understand each step of the process. However, the complexity of the code might limit its reusability. A simpler approach like the one in the example solution might be more reusable.

Here's a suggestion for a simpler version of your query that directly answers the question:

```sql
SELECT player_name, team_abbreviation, SUM(pts) as total_points
FROM bootcamp.nba_game_details
GROUP BY player_name, team_abbreviation
ORDER BY total_points DESC
LIMIT 1
```

This query simply groups the data by player and team, sums the points for each group, and then orders the results by the total points in descending order. It then returns the player and team with the highest total points.

### Feedback for `query_4.sql`

The student's solution is quite comprehensive and shows a good understanding of SQL, especially the use of `WITH` clauses, `JOIN` operations, and `GROUP BY GROUPING SETS`. However, there are a few areas where the solution could be improved.

1) **Coverage of the homework prompt:** The student's solution does not directly answer the question in the prompt. The prompt asks for the player who scored the most points in one season, but the student's solution calculates the average points (`avg_pts`) and total points (`sum_pts`) for each player in each season and for each team. The student should modify the query to focus on the total points per player per season.

2) **Logical coherence:** The student's solution is logically coherent, but it is more complex than necessary. The student first deduplicates the `nba_game_details` table, then joins it with the `nba_games` table to get the season, and finally uses `GROUP BY GROUPING SETS` to calculate the average and total points. However, the prompt does not require deduplication or calculation of average points. The student could simplify the solution by directly joining the `nba_game_details` and `nba_games` tables and using a simple `GROUP BY` to calculate the total points per player per season.

3) **Readability:** The student's solution is well-structured and includes comments explaining each step, which is good practice. However, the variable names could be more descriptive. For example, `olap_cube` does not clearly convey the purpose of the variable. A more descriptive name like `player_season_points` would improve readability.

4) **Robustness and error handling:** The student's solution does not include any error handling. The student could add checks to ensure that the data in the `nba_game_details` and `nba_games` tables is valid and complete before performing the calculations.

5) **Modularity and reusability:** The student's solution is modular and uses `WITH` clauses to break down the problem into smaller parts. However, the solution is not very reusable because it is tailored to the specific question in the prompt. The student could improve reusability by defining a general function to calculate the total points per player per season, which could be reused in other queries.

Here is a simplified version of the student's solution that directly answers the question in the prompt:

```sql
WITH player_season_points AS (
  SELECT
    gd.player_name,
    g.season,
    SUM(gd.pts) AS total_points
  FROM bootcamp.nba_game_details gd
    JOIN bootcamp.nba_games g ON gd.game_id = g.game_id AND gd.team_id = g.home_team_id
  GROUP BY gd.player_name, g.season
)
SELECT
  player_name,
  season,
  total_points
FROM player_season_points
ORDER BY total_points DESC
LIMIT 1;
```

This query calculates the total points per player per season and returns the player and season with the most points.

### Feedback for `query_5.sql`

The student's solution is a good attempt at solving the problem, but there are a few areas that could be improved. Here's my feedback:

1) **Coverage of the Homework Prompt:** The student's solution does not fully cover the homework prompt. The prompt asks to build additional queries on top of the results of the `GROUPING SETS` aggregations, but the student's solution does not use `GROUPING SETS`. Instead, it uses a simple `GROUP BY` clause. The student should revise their solution to incorporate the use of `GROUPING SETS` as per the assignment prompt.

2) **Logical Flow:** The student's solution logically follows the question asked, "Which team has won the most games?". However, it does not follow the instructions given in the homework prompt. The student should ensure that they follow all instructions in the prompt, even if they believe a simpler solution is possible.

3) **Readability:** The student's code is well-formatted and easy to read. The use of comments is commendable as it helps to understand the thought process behind the code. However, the comment "This question doesn't require a GROUPING SETS query" contradicts the assignment prompt, which specifically asks for a `GROUPING SETS` query.

4) **Robustness and Error Handling:** The student's code does not handle the case where the game is won by the away team. The code only considers games won by the home team. This could lead to incorrect results. The student should revise their code to consider games won by both home and away teams.

5) **Modularity and Reusability:** The student's code is modular and could be reused in other contexts with minor modifications. However, the code could be made more reusable by creating a function or stored procedure that takes the team id as a parameter and returns the number of games won by that team.

In conclusion, the student should revise their solution to incorporate the use of `GROUPING SETS` as per the assignment prompt and ensure that games won by both home and away teams are considered.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly implemented the state change tracking for `nba_players` and your use of the `CASE` statement to determine the player's state is appropriate and aligns with the requirements of the prompt. 

However, there are a few areas where your solution could be improved:

1) **Covering all aspects of the prompt:** Your solution does not seem to handle the tracking of first and last active seasons for each player, which is a requirement in the example solution. You might want to add `first_active_season` and `last_active_season` fields to your table and update them accordingly.

2) **Readability:** Your code is generally well-formatted and easy to read. However, it would be beneficial to include more comments explaining your thought process, especially for the complex `CASE` statement. This would make it easier for others (and future you) to understand your code.

3) **Robustness and error handling:** Your solution assumes that the `seasons` array is sorted and that the last element corresponds to the last active season. If this is not the case, your `years_since_last_active` calculation might be incorrect. You should add some error handling or checks to ensure that the data is in the expected format.

4) **Modularity and reusability:** Your solution is specific to the `nba_players` table and the current season. If you wanted to use this code for a different table or season, you would need to make several changes. Consider parameterizing your query or turning it into a stored procedure to increase its reusability.

Here's a suggestion on how you could modify your code to address these points:

```sql
CREATE TABLE nba_players_state_tracking AS
SELECT
  player_name,
  seasons,
  is_active,
  current_season - MAX(seasons) AS years_since_last_active,
  current_season,
  CASE
    WHEN is_active AND CARDINALITY(seasons) = 1 THEN 'New'
    WHEN NOT is_active AND current_season - MAX(seasons) = 1 THEN 'Retired'
    WHEN is_active AND CARDINALITY(seasons) > 1 AND current_season - MAX(seasons) = 0 THEN 'Continued Playing'
    WHEN is_active AND current_season - MAX(seasons) > 1 THEN 'Returned from Retirement'
    WHEN NOT is_active AND current_season - MAX(seasons) > 1 THEN 'Stayed Retired'
  END AS player_state
FROM bootcamp.nba_players
GROUP BY player_name, seasons, is_active, current_season
```

In this version, I've added a `GROUP BY` clause to calculate `years_since_last_active` based on the maximum season in the `seasons` array, which should be more robust. I've also updated the `CASE` statement to use this new calculation.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to perform aggregations along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the homework prompt. The prompt asks for the aggregation of `game_details` data along the dimensions of player and team, player and season, and team. However, it also requires the calculation of total player points and total team wins. Your solution correctly calculates the total player points (`sum_pts`), but it does not calculate the total team wins. You can add this calculation to your query by following the example solution.

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments explaining your thought process, which is excellent. However, the alias `olap_cube` for the final subquery might be a bit confusing, as it doesn't accurately represent the data being returned. Consider renaming it to something more descriptive, like `aggregated_data`.

3) **Robustness and error handling:** Your code does not handle the case where the `team_id` in the `nba_game_details` table corresponds to the `visitor_team_id` in the `nba_games` table. In the example solution, this is handled by a `CASE` statement that determines whether the team won or lost based on the `home_team_wins` column and the `team_id`. You should consider adding similar logic to your solution.

4) **Modularity and reusability:** Your code is modular and well-structured, with each subquery performing a distinct operation. This makes it easy to understand and reuse specific components in other contexts. However, the `combined` subquery could be made more reusable by removing the `gd.row_num = 1` condition from the `WHERE` clause and adding it to the `JOIN` condition instead. This would allow the subquery to be used in other contexts where it might be necessary to join on rows other than the first one.

Here's how you could modify your code to address these points:

```sql
WITH nba_game_details_deduped AS (
  SELECT
    game_id,
    team_id,
    team_abbreviation,
    player_id,
    player_name,
    pts,
    ROW_NUMBER() OVER (PARTITION BY game_id, team_id, player_id ORDER BY pts DESC) AS row_num
  FROM bootcamp.nba_game_details
),
combined AS (
  SELECT
    gd.game_id,
    gd.team_abbreviation,
    gd.player_name,
    gd.pts,
    g.season,
    CASE 
      WHEN gd.team_id = g.visitor_team_id AND g.home_team_wins = 1 THEN 0
      WHEN gd.team_id = g.visitor_team_id AND g.home_team_wins = 0 THEN 1
      ELSE g.home_team_wins
    END AS won
  FROM nba_game_details_deduped gd
    JOIN bootcamp.nba_games g ON gd.game_id = g.game_id AND gd.row_num = 1
),
aggregated_data AS (
  SELECT
    COALESCE(CAST(team_abbreviation AS VARCHAR), '(overall)') AS team,
    COALESCE(CAST(player_name AS VARCHAR), '(overall)') AS player,
    COALESCE(CAST(season AS VARCHAR), '(overall)') AS season,
    SUM(pts) AS sum_pts,
    SUM(won) AS total_wins
  FROM combined
  GROUP BY GROUPING SETS (
    (player_name, team_abbreviation),
    (player_name, season),
    (team_abbreviation)
  )
)
SELECT
  *
FROM
  aggregated_data
```

Keep up the good work! With a few adjustments, your solution will fully meet the requirements of the prompt.

### Feedback for `query_3.sql`

Your solution is quite comprehensive and it's clear that you've put a lot of thought into it. You've successfully answered the question in the prompt, which is to find the player who scored the most points playing for a single team. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the homework prompt:** Your solution goes beyond the requirements of the prompt by including a season in your grouping sets. While this is not incorrect, it's not necessary for answering the question and makes your solution more complex than it needs to be. The example solution focuses only on the player and team, which is sufficient for this task.

2) **Logical structure and coherence:** Your solution is logically structured and coherent. You've clearly explained your thought process in your comments, which is very helpful. However, the complexity of your solution could be reduced. For example, the deduplication of the `nba_game_details` table might not be necessary if the source data is clean. If you have reason to believe that there are duplicates in the source data, it would be helpful to include a comment explaining this.

3) **Readability:** Your code is generally well-formatted and easy to read. Your use of meaningful variable names and comments is excellent. However, the complexity of your solution makes it a bit harder to follow than the example solution. Simplifying your solution would improve its readability.

4) **Robustness and error handling:** Your solution seems robust and should handle a variety of inputs well. However, it's not clear how your solution would handle null values. The example solution includes a check for null values in the `total_player_points` column, which is a good practice to follow.

5) **Modularity and reusability:** Your solution is modular and could be reused in other contexts with some modifications. However, the complexity of your solution might make it harder to adapt for other tasks. The example solution is simpler and more focused, which makes it more reusable.

In summary, your solution is correct and well-structured, but could be simplified. I would suggest removing the unnecessary parts, such as the deduplication of the `nba_game_details` table and the inclusion of the season in your grouping sets. This would make your solution easier to read and understand, and more reusable in other contexts.

### Feedback for `query_4.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You've successfully used `GROUPING SETS` to answer the question, "Which player scored the most points in one season?" However, there are a few areas where your solution could be improved.

1) **Coverage of the Homework Prompt:** Your solution covers the main aspects of the homework prompt, but it seems to be more complex than necessary. The prompt asks for a query to find the player who scored the most points in a single season. Your solution, however, includes additional grouping sets that aren't required to answer this question. For example, the grouping set `(player_name, team_abbreviation)` and `(team_abbreviation)` are not necessary for this specific query.

2) **Readability:** Your code is generally well-structured and includes helpful comments, which is great. However, the complexity of your solution makes it a bit harder to read and understand. Simplifying your solution to only include the necessary grouping sets would improve readability.

3) **Robustness and Error Handling:** Your solution includes a deduplication step, which is a good practice for ensuring the robustness of your results. However, it's not clear from the prompt whether this step is necessary. If the source data is expected to be clean, this step might not be needed.

4) **Modularity and Reusability:** Your solution is modular and includes several subqueries that could be reused in other contexts. However, the `olap_cube` subquery includes unnecessary grouping sets, which could make it less useful for other queries.

Here's a simplified version of your solution that only includes the necessary grouping set:

```sql
WITH combined AS (
  SELECT
    gd.player_name,
    g.season,
    gd.pts
  FROM bootcamp.nba_game_details gd
    JOIN bootcamp.nba_games g ON gd.game_id = g.game_id AND gd.team_id = g.home_team_id
),
grouped AS (
  SELECT
    player_name,
    season,
    SUM(pts) AS sum_pts
  FROM combined
  GROUP BY player_name, season
)
SELECT
  *
FROM
  grouped
ORDER BY
  sum_pts DESC
LIMIT 1
```

This version of the query is simpler and easier to read, and it directly answers the question in the prompt.

### Feedback for `query_5.sql`

Overall, your solution is logically sound and it does answer the question in the prompt. However, there are a few areas where it could be improved to better align with the prompt and to enhance readability and robustness.

1) **Alignment with the prompt:** The prompt asks you to build additional queries on top of the results of the `GROUPING SETS` aggregations. Your solution, while correct, does not use the `GROUPING SETS` results. Instead, it directly queries the `nba_games` table. To better align with the prompt, you could modify your query to use the `game_details_dashboard` table, which presumably contains the results of the `GROUPING SETS` aggregations.

2) **Readability:** Your code is generally well-structured and easy to read. You've used meaningful variable names and included comments to explain your thought process, which is great. However, the comment "Filtering on home_team_wins = 1 still brings all games while ensuring that one record per game is returned" might be a bit confusing. It would be clearer to say something like "Filtering on home_team_wins = 1 ensures that we only count games that the team won."

3) **Robustness and error handling:** Your code does not handle the case where there might be a tie for the most games won. If two teams have won the same number of games, your query will only return one of them. To handle this case, you could remove the `LIMIT 1` clause and instead use a subquery to filter for the maximum number of games won.

4) **Modularity and reusability:** Your code is modular and could be easily adapted to answer similar questions. However, to improve reusability, you could consider parameterizing the team_id, so that the query could be used to find the number of games won by any given team.

Here's an example of how you might modify your query to address these points:

```sql
-- Which team has won the most games?
WITH winning_teams AS (
  SELECT
    team_name,
    total_team_wins
  FROM game_details_dashboard
  WHERE aggregation_level = 'player_name__team_name'
)
SELECT
  team_name,
  total_team_wins AS games_won
FROM
  winning_teams
WHERE
  total_team_wins = (SELECT MAX(total_team_wins) FROM winning_teams)
ORDER BY
  total_team_wins DESC;
```

Keep up the good work, and remember to closely follow the prompt's instructions.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a given 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution does not fully cover the prompt. The prompt asks for the most games won by each team in a 90-game stretch, but your solution only returns the team with the highest number of wins in a 90-game stretch. To cover the full prompt, you should remove the `LIMIT 1` clause from your final query and group by `team_abbreviation` to get the maximum `rolling_90_day_wins` for each team.

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process. However, the comment for the `cumulative_wins` calculation in the `combined` CTE is slightly misleading. It suggests that you are calculating cumulative wins, but you are actually calculating the win status for each game. A more accurate comment might be "-- Calculate win status for each game for each team".

3) **Robustness and error handling:** Your code does not handle the case where there are fewer than 90 games for a team. In this case, the `LAG` function will return NULL, and the subtraction operation will also return NULL. You should handle this case by using the `COALESCE` function to replace NULL with 0.

4) **Modularity and reusability:** Your code is modular and the use of CTEs makes it easy to understand and reuse specific components. However, the `combined` and `cumulated_wins` CTEs are very similar and could be combined into a single CTE to simplify the code.

Here is a revised version of your code with these suggestions:

```sql
-- CTE to dedupe and retrieve the games for each team
WITH nba_game_details_deduped AS (
	SELECT DISTINCT
        game_id,
		team_id,
		team_abbreviation
	FROM
        bootcamp.nba_game_details
),
combined AS (
	SELECT
        gd.team_id,
		gd.game_id,
		gd.team_abbreviation,
		g.game_date_est,
        -- Calculate win status for each game for each team
		SUM(CASE WHEN gd.team_id = g.team_id_home THEN g.home_team_wins ELSE 1 - g.home_team_wins END) OVER (PARTITION BY gd.team_id, gd.team_abbreviation ORDER BY g.game_date_est) AS cumulative_wins,
        -- Calculate rolling 90 game wins
        SUM(CASE WHEN gd.team_id = g.team_id_home THEN g.home_team_wins ELSE 1 - g.home_team_wins END) OVER (PARTITION BY gd.team_id, gd.team_abbreviation ORDER BY g.game_date_est ROWS BETWEEN 89 PRECEDING AND CURRENT ROW) - COALESCE(LAG(SUM(CASE WHEN gd.team_id = g.team_id_home THEN g.home_team_wins ELSE 1 - g.home_team_wins END) OVER (PARTITION BY gd.team_id, gd.team_abbreviation ORDER BY g.game_date_est ROWS BETWEEN 89 PRECEDING AND CURRENT ROW), 90, 0) OVER (PARTITION BY gd.team_id, gd.team_abbreviation ORDER BY g.game_date_est), 0) AS rolling_90_day_wins
	FROM
        nba_game_details_deduped gd JOIN bootcamp.nba_games g ON gd.game_id = g.game_id
)
SELECT
    team_abbreviation,
	MAX(rolling_90_day_wins) AS max_rolling_90_day_wins
FROM
    combined
GROUP BY
    team_abbreviation
ORDER BY
    max_rolling_90_day_wins DESC
```

Keep up the good work! Your understanding of SQL and window functions is clear from your solution. With a few tweaks, your code will be even better.

### Feedback for `query_7.sql`

Overall, your solution is well-structured and logically follows the homework prompt. You have correctly used window functions to answer the question about LeBron James' scoring streak. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution covers the main aspects of the homework prompt, but it seems to be missing the deduplication step that is present in the example solution. The example solution uses `ROW_NUMBER() OVER (PARTITION BY gd.game_id, team_id, player_id ORDER BY g.game_date_est)` to ensure that there are no duplicate entries for the same game, team, and player. If there are duplicates in your data, your current solution might overcount the number of games where LeBron James scored over 10 points.

2) **Readability:** Your code is generally well-formatted and easy to read. You've used meaningful names for your Common Table Expressions (CTEs) and variables, which makes it easier to understand what each part of your code is doing. However, you could improve the readability of your code by adding more comments to explain your thought process, especially for the more complex parts of your code like the window functions.

3) **Robustness and error handling:** Your code does not currently handle potential errors or edge cases. For example, what happens if the `player_name` is not found in the database, or if there are no games where LeBron James scored over 10 points? Adding some error handling or checks for these situations could make your code more robust.

4) **Modularity and reusability:** Your use of CTEs makes your code more modular and easier to understand. However, the query is specific to LeBron James and games where he scored over 10 points. If you wanted to reuse this code to analyze other players or different point thresholds, you would need to manually change these values in your code. Consider parameterizing your query to make it more reusable.

Here's how you might modify your code to address these points:

```sql
-- Parameterized player name and points threshold
DECLARE @player_name varchar(50) = 'LeBron James';
DECLARE @points_threshold int = 10;

WITH deduped_games AS (
    SELECT DISTINCT
        g.game_date_est,
        CASE
            WHEN gd.pts > @points_threshold THEN 1
            ELSE 0
        END AS scored_over_threshold,
        ROW_NUMBER() OVER (PARTITION BY gd.game_id, team_id, player_id ORDER BY g.game_date_est) AS row_num_dedup
    FROM
        bootcamp.nba_game_details gd
        JOIN bootcamp.nba_games g ON gd.game_id = g.game_id
    WHERE
        player_name = @player_name
        AND row_num_dedup = 1
),
streaks AS (
    SELECT
        game_date_est,
        scored_over_threshold,
        ROW_NUMBER() OVER (ORDER BY game_date_est) - 
        ROW_NUMBER() OVER (PARTITION BY scored_over_threshold ORDER BY game_date_est) AS streak_identifier
    FROM
        deduped_games
)
SELECT
    COUNT(1) AS number_of_games
FROM
    streaks
WHERE
    scored_over_threshold = 1
GROUP BY
    streak_identifier
ORDER BY
    number_of_games DESC
LIMIT 1;
```

This version of the code deduplicates the games, handles the case where the player name is not found or there are no games where the player scored over the threshold, and makes it easy to change the player name and points threshold.

**Review by prajapatipreeti — APPROVED** [review]
